### ISSUE TITLE
"Could not load //external package" error when --incompatible_no_transitive_loads option is used

### ISSUE THREAD

**Issue by davido** [bazelbuild/bazel#6539]
I'm trying to build Gerrit master on most recent Bazel version: 4ca2b6d5c4d201009ff21a3abbeddd9d58d4c95b.

All is fine when buildng gerrit `headless` server without `--all_incompatible_changes` option:

```shell
  $ bazel build :headless
INFO: Invocation ID: 367db4ba-df05-408d-8b7d-1e8da51459be
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:98:7: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:99:7: Make sure that you are running at least Bazel 0.17.1.
INFO: Analysed target //:headless (223 packages loaded, 4223 targets configured).
INFO: Found 1 target...
Target //:headless up-to-date:
  bazel-bin/headless.war
INFO: Elapsed time: 11.507s, Critical Path: 1.75s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
```

However, rien ne va plus, if `--all_incompatible_changes` option is used:

```shell
  $ bazel build --all_incompatible_changes :headless
Starting local Bazel server and connecting to it...
INFO: Invocation ID: c844219d-d777-46e5-ac1d-beafb7d0fbcd
ERROR: /home/davido/projects/gerrit2/WORKSPACE:33:1: file '@bazel_skylib//:lib.bzl' does not contain symbol 'versions'
ERROR: /home/davido/projects/gerrit2/WORKSPACE:35:1: name 'versions' is not defined
ERROR: Error evaluating WORKSPACE file
ERROR: error loading package '': Encountered error while reading extension file 'go/def.bzl': no such package '@io_bazel_rules_go//go': error loading package 'external': Could not load //external package
ERROR: error loading package '': Encountered error while reading extension file 'go/def.bzl': no such package '@io_bazel_rules_go//go': error loading package 'external': Could not load //external package
INFO: Elapsed time: 2.712s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

**Comment by davido** [comment]
I tracked this down to `--incompatible_no_transitive_loads` option, thus related to: #5636.

**Comment by laurentlb** [comment]
This seems to work as intended for Bazel.

Recommendation is to stop using `lib.bzl` in skylib. It was recently deprecated: https://github.com/bazelbuild/bazel-skylib/blob/master/lib.bzl#L17
https://github.com/bazelbuild/bazel-skylib/pull/58/files

Try loading `lib/versions.bzl` instead.

**Comment by davido** [comment]
It helped with `skylib`:

```diff
diff --git a/WORKSPACE b/WORKSPACE
index 37080b780b..85618500b7 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -30,7 +30,7 @@ http_file(
     urls = ["https://raw.githubusercontent.com/google/closure-compiler/35d2b3340ff23a69441f10fa3bc820691c2942f2/contrib/externs/polymer-1.0.js"],
 )
 
-load("@bazel_skylib//:lib.bzl", "versions")
+load("@bazel_skylib//lib:versions.bzl", "versions")
 
 versions.check(minimum_bazel_version = "0.17.1")

```

However, gerrit build is still failing in `rules_clsoure` and `go-lang`:

```
$ bazel build --incompatible_no_transitive_loads :headless 
ERROR: /home/davido/projects/gerrit2/WORKSPACE:37:1: file '@io_bazel_rules_closure//closure:defs.bzl' does not contain symbol 'closure_repositories'
ERROR: /home/davido/projects/gerrit2/WORKSPACE:43:1: name 'closure_repositories' is not defined
ERROR: Error evaluating WORKSPACE file
ERROR: error loading package '': Encountered error while reading extension file 'go/def.bzl': no such package '@io_bazel_rules_go//go': error loading package 'external': Could not load //external package
ERROR: error loading package '': Encountered error while reading extension file 'go/def.bzl': no such package '@io_bazel_rules_go//go': error loading package 'external': Could not load //external package
INFO: Elapsed time: 0.192s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

**Comment by laurentlb** [comment]
Thanks, I've filed https://github.com/bazelbuild/rules_closure/issues/305

**Comment by solstice333** [comment]
I ran into this problem when trying to use a bazel 1.1.0 to build a plugin for gerrit 2.15.7. I wanted to avoid the scenario of me trying to bump the versions of possibly multiple dependencies and risk other things breaking to achieve the more correct explicit re-export. I ended up using `bazel build --verbose_failures --incompatible_no_transitive_loads=false plugins/<name of plugin>` (the default for --incompatible_no_transitive_loads is **true** per the doc).

**Comment by davido** [comment]
@solstice333 Note, that the latest Gerrit release on 2.15 release line is 2.15.17. You should always use the latest fix release. And then Bazel 1.1.0 should just work.

**Comment by solstice333** [comment]
@davido 

> However, in case you cannot bump Gerrit version to 2.15.17 and really really have to use outdated 2.15.7

Yeah unfortunately that's the current preferred scenario since I ended up forking the gerrit repo and making a few small hacks to get a tree-driven plugins/hooks build to work under bazel 0.22.0 on Windows. Also ended up making some tooling to automate the installation steps specified in the 2.15.7 docs. But all that is probably outside the scope of this thread, so I won't go into that here. I do plan to make an effort to move to the latest release and see how that goes.

> then just check the pinned Bazel version in WORKSPACE file

Thanks for this tip. I'll keep that in mind.

> Note, though, that in latest release we added .bazelversion file, and using bazelisk it should just work

Oh that's neat. I'll also keep that in mind.

**Comment by davido** [comment]
> I do plan to make an effort to move to the latest release and see how that goes.

Note, that there is a fundamental difference between upgrade to the latest releases with bump the fix release level.

To upgrade from 2.15.x to 2.16.x and then to 3.0.x and upcoming 3.1.0 would take substantial efforts. Whereas bump from 2.15.7 to 2.15.7 should ab as trivial as stop gerrit replace gerrit.war with new version and re-start gerit. C'est tout.

**Comment by solstice333** [comment]
Ah right, that makes sense. I feel like I would also need to make sure the plugins are rebuilt on the tag that matches the full release name (in other words, if I bump my release level to 2.15.17, then I'd build my plugins at tag 2.15.17). Is this true?

**Comment by davido** [comment]
> Is this true?

Correct. However given that the plugin API should stay stable between 2.15.7 and 2.15.17 I wouldn't expect any surprizes here.

So, just replace the core plugins from release.war and re-build your custom plugins, done.

**Comment by solstice333** [comment]
Got it. Ok, sounds straightforward enough. Thanks for your help!

**Comment by solstice333** [comment]
Well, took me awhile longer than I thought it would. It was partly user error as I have multiple Visual Studio installations. The Bazel installation doc clearly mentions to set `BAZEL_VC` if this is the case. I set it to my VS2019/VC14 install and the build got passed protobuf, yay. The next issue I ran into is that I have git-bash for Windows installed. There's a file in the gerrit repo `tools/workspace-status.sh` that I expected Bazel to execute using `C:\msys64\usr\bin\bash.exe` b.c. that's what I had `BAZEL_SH` set to per the Bazel installation doc, except it didn't, and ran it with git-bash instead. I ended up going to the properties of that .sh and telling it to open with `C:\msys64\usr\bin\bash.exe` instead, and that worked. Both the plugins I care about built successfully, whew. I'll open up an issue for the `BAZEL_SH` problem that I think is probably a bazel bug.
